pkg/virtualfs/directory.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2020 The Kanister Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package virtualfs
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/kopia/kopia/fs"
+	"github.com/pkg/errors"
+)
+
+// Directory is a mock in-memory implementation of kopia's fs.Directory
+type Directory struct {
+	dirEntry
+
+	children     fs.Entries
+	readdirError error
+	onReaddir    func()
+}
+
+// Summary returns summary of a directory.
+func (d *Directory) Summary() *fs.DirectorySummary {
+	return nil
+}
+
+// AddFileWithStreamSource adds a mock file with the specified name, permissions and source.
+func (d *Directory) AddFileWithStreamSource(name, sourceEndpoint string, permissions os.FileMode) (file *File, err error) {
+	d, name, err = d.resolveAndAddSubdir(name)
+	if err != nil {
+		return nil, err
+	}
+
+	file = &File{
+		dirEntry: dirEntry{
+			name: name,
+			mode: permissions,
+			// TODO: Add owner information
+		},
+		source: func() (ReaderSeekerCloser, error) {
+			return streamReader(sourceEndpoint)
+		},
+	}
+
+	err = d.addChild(file)
+
+	return file, errors.Wrap(err, "Failed to add file")
+}
+
+// AddDir adds a fake directory with a given name and permissions.
+func (d *Directory) AddDir(name string, permissions os.FileMode) (subdir *Directory, err error) {
+	d, name, err = d.resolveSubdir(name)
+	if err != nil {
+		return nil, err
+	}
+
+	subdir = &Directory{
+		dirEntry: dirEntry{
+			name: name,
+			mode: permissions | os.ModeDir,
+		},
+	}
+
+	err = d.addChild(subdir)
+
+	return subdir, err
+}
+
+// Subdir finds a subdirectory with the given name.
```

Comment:
Is there an intended use for `Subdir` or simply copied over? If it is not needed, let's remove it.